### PR TITLE
refactor: merge files to save lines of code

### PR DESCRIPTION
**Pull request by MerlinEgalite**

### CODE REVIEW

**Comment by Jean-Grimal** [comment]
I think I would rather keep the two files

**Comment by MerlinEgalite** [comment]
> I think I would rather keep the two files

Why? I think it's unnecessary since for the following reasons:
- It's adding more LOC 
- We're not going to reuse any of those 2 files in other protocols

File: src/ERC20DelegatesUpgradeable.sol [hunk 1]
```diff
@@ -1,38 +1,151 @@
...
+    function _getVotingUnits(address account) internal view returns (uint256) {
         return balanceOf(account);
     }
+
+    /// @dev Move voting power when tokens are transferred.
+    ///
+    /// Emits a {IDelegates-DelegateVotesChanged} event.
+    function _update(address from, address to, uint256 value) internal virtual override {
+        super._update(from, to, value);
+        // No check of supply cap here like in OZ implementation has MORPHO has a 1B total supply cap.
```
Comment (MerlinEgalite) [inline src/ERC20DelegatesUpgradeable.sol:148]:
```suggestion
        // No check of supply cap here like in OZ implementation as MORPHO has a 1B total supply cap.
```

File: src/ERC20DelegatesUpgradeable.sol [hunk 2]
```diff
@@ -1,38 +1,151 @@
...
+    /// @custom:storage-location erc7201:morpho.storage.ERC20Delegates
+    struct ERC20DelegatesStorage {
+        mapping(address account => address) _delegatee;
+        mapping(address delegatee => uint256) _votingPower;
+        uint256 _totalVotingPower;
+    }
+
+    /* FUNCTIONS */
+
+    function _getDelegatesStorage() private pure returns (ERC20DelegatesStorage storage $) {
```
Comment (Jean-Grimal) [inline src/ERC20DelegatesUpgradeable.sol:49]:
```suggestion
    function _getERC20DelegatesStorage() private pure returns (ERC20DelegatesStorage storage $) {
```

File: src/ERC20DelegatesUpgradeable.sol [hunk 3]
```diff
@@ -1,38 +1,151 @@
...
+        assembly {
+            $.slot := ERC20DelegatesStorageLocation
+        }
+    }
+
+    /// @dev Returns the current amount of votes that `account` has.
+    function getVotes(address account) public view virtual returns (uint256) {
+        ERC20DelegatesStorage storage $ = _getDelegatesStorage();
```
Comment (Jean-Grimal) [inline src/ERC20DelegatesUpgradeable.sol:57]:
```suggestion
        ERC20DelegatesStorage storage $ = _getERC20DelegatesStorage();
```

File: src/ERC20DelegatesUpgradeable.sol [hunk 4]
```diff
@@ -1,38 +1,151 @@
...
+        return $._votingPower[account];
+    }
+
+    /// @dev Returns the current total supply of votes.
+    function _getTotalSupply() internal view virtual returns (uint256) {
+        ERC20DelegatesStorage storage $ = _getDelegatesStorage();
```
Comment (Jean-Grimal) [inline src/ERC20DelegatesUpgradeable.sol:63]:
```suggestion
        ERC20DelegatesStorage storage $ = _getERC20DelegatesStorage();
```

File: src/ERC20DelegatesUpgradeable.sol [hunk 5]
```diff
@@ -1,38 +1,151 @@
...
-    /**
-     * @dev Returns the voting units of an `account`.
-     *
-     * WARNING: Overriding this function may compromise the internal vote accounting.
-     * `ERC20Delegates` assumes tokens map to voting units 1:1 and this is not easy to change.
-     */
-    function _getVotingUnits(address account) internal view virtual override returns (uint256) {
+    /// @dev Returns the delegate that `account` has chosen.
+    function delegates(address account) public view virtual returns (address) {
+        ERC20DelegatesStorage storage $ = _getDelegatesStorage();
```
Comment (Jean-Grimal) [inline src/ERC20DelegatesUpgradeable.sol:69]:
```suggestion
        ERC20DelegatesStorage storage $ = _getERC20DelegatesStorage();
```

File: src/ERC20DelegatesUpgradeable.sol [hunk 6]
```diff
@@ -1,38 +1,151 @@
...
+        $._delegatee[account] = delegatee;
+
+        emit DelegateChanged(account, oldDelegate, delegatee);
+        _moveDelegateVotes(oldDelegate, delegatee, _getVotingUnits(account));
+    }
+
+    /// @dev Transfers, mints, or burns voting units. To register a mint, `from` should be zero. To register a burn, `to`
+    /// should be zero. Total supply of voting units will be adjusted with mints and burns.
+    function _transferVotingUnits(address from, address to, uint256 amount) internal virtual {
+        ERC20DelegatesStorage storage $ = _getDelegatesStorage();
```
Comment (Jean-Grimal) [inline src/ERC20DelegatesUpgradeable.sol:109]:
```suggestion
        ERC20DelegatesStorage storage $ = _getERC20DelegatesStorage();
```

File: src/ERC20DelegatesUpgradeable.sol [hunk 7]
```diff
@@ -1,38 +1,151 @@
...
+        }
+        if (to == address(0)) {
+            $._totalVotingPower -= amount;
+        }
+        _moveDelegateVotes(delegates(from), delegates(to), amount);
+    }
+
+    /// @dev Moves delegated votes from one delegate to another.
+    function _moveDelegateVotes(address from, address to, uint256 amount) private {
+        ERC20DelegatesStorage storage $ = _getDelegatesStorage();
```
Comment (Jean-Grimal) [inline src/ERC20DelegatesUpgradeable.sol:121]:
```suggestion
        ERC20DelegatesStorage storage $ = _getERC20DelegatesStorage();
```

File: src/ERC20DelegatesUpgradeable.sol [hunk 8]
```diff
@@ -1,38 +1,151 @@
...
+        );
+        _useCheckedNonce(signer, nonce);
+        _delegate(signer, delegatee);
+    }
+
+    /// @dev Delegate all of `account`'s voting units to `delegatee`.
+    ///
+    /// Emits events {IDelegates-DelegateChanged} and {IDelegates-DelegateVotesChanged}.
+    function _delegate(address account, address delegatee) internal virtual {
+        ERC20DelegatesStorage storage $ = _getDelegatesStorage();
```
Comment (Jean-Grimal) [inline src/ERC20DelegatesUpgradeable.sol:98]:
```suggestion
        ERC20DelegatesStorage storage $ = _getERC20DelegatesStorage();
```

**Review by Jean-Grimal — CHANGES_REQUESTED** [review]

**Pull request merged**